### PR TITLE
[SPARK-38320][SS] Fix flatMapGroupsWithState timeout in batch with data for key

### PR DESCRIPTION
**Pull request by alex-balikov**
### What changes were proposed in this pull request?

As described in [SPARK-38320](https://issues.apache.org/jira/browse/SPARK-38320), the bug is that it is possible for (flat)MapGroupsWithState to timeout a key even if that key received data within the same batch. This is against the documented (flat)MapGroupsWithState contract. The problem is due to the StateStore.iterator not reflecting StateStore changes made *after* its creation - this is illustrated in the test this PR adds to StateStoreSuite.scala.

The fix is to *late bind* the timeoutProcessorIter timeout processing iterator in FlatMapGroupsWithStateExec to be created *after* the input iterator has been exhausted and the state changes applied to the StateStore.

### Why are the changes needed?

The changes are needed to ensure the state timeout processing iterator for (flat)MapGroupsWithState is created *after* the input is processed and the changes are applied into the StateStore, otherwise it may not notice these changes (the change to the key timeout timestamp being updated as part of the input processing).

### Does this PR introduce _any_ user-facing change?

No. Bug fix.

### How was this patch tested?

* Added a test to StateStoreSuite.scala to illustrate the difference of state store iterator behavior across the different implementations of state stores. In particular the test illustrates the RocksDB state store iterator not reflecting state store changes made after its creation.
* Added test to FlatMapGroupsWithStateSuite.scala which would fail with unexpected state timeout if the issue was not fixed.

### CODE REVIEW

**Comment by alex-balikov** [comment]
@HeartSaVioR can you please review. Thanks

**Comment by HeartSaVioR** [comment]
Thanks for pinging me. I'll review this today. 
cc. @tdas @zsxwing @viirya @xuanyuanking for seeking another eyes of reviews

**Review by HeartSaVioR — APPROVED** [review]
+1 Thanks for investigating such tricky issue and fixing it properly!

**Comment by HeartSaVioR** [comment]
I'll keep this open for today to see further feedback. I'll merge if there is no outstanding comment in tomorrow.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala [hunk 1]
```diff
@@ -172,12 +172,20 @@ case class FlatMapGroupsWithStateExec(
           timeoutProcessingStartTimeNs = System.nanoTime
         })
 
-    val timeoutProcessorIter =
-      CompletionIterator[InternalRow, Iterator[InternalRow]](processor.processTimedOutState(), {
-        // Note: `timeoutLatencyMs` also includes the time the parent operator took for
-        // processing output returned through iterator.
-        timeoutLatencyMs += NANOSECONDS.toMillis(System.nanoTime - timeoutProcessingStartTimeNs)
-      })
+    // SPARK-38204: Late-bind the timeout processing iterator so it is created *after* the input is
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala:175]: SPARK-38204 or SPARK-38320?

**Review by viirya — APPROVED** [review]
lgtm

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala:175]: Nice finding :)
@alex-balikov looks like JIRA ticket is not properly mentioned here.

Comment (alex-balikov) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala]: Thanks for the catch. Fixed.

**Comment by HeartSaVioR** [comment]
I don't see any outstanding unresolved comment now. Thanks! Merging to master.

**Comment by HeartSaVioR** [comment]
Thanks @alex-balikov for the contribution! I merged into master/3.2.

**Pull request closed without merging**